src/red_black_tree.jl

```diff
@@ -0,0 +1,380 @@
+# it has unique keys
+# leftChild has keys which are less than the node
+# rightChild has keys which are greater than the node
+# color is true if it's a Red Node, else it's false
+mutable struct RBTreeNode{K}
+    color::Bool 
+    data::Union{K, Nothing}
+    leftChild::Union{Nothing, RBTreeNode{K}}
+    rightChild::Union{Nothing, RBTreeNode{K}}
+    parent::Union{Nothing, RBTreeNode{K}}
+
+    RBTreeNode{K}() where K = new{K}(true, nothing, nothing, nothing, nothing)
+    
+    RBTreeNode{K}(d::K) where K = new{K}(true, d, nothing, nothing, nothing)
+end
+
+RBTreeNode() = RBTreeNode{Any}()
+RBTreeNode(d) = RBTreeNode{Any}(d)
+
+function Base.isequal(x::RBTreeNode{K}, y::RBTreeNode{K})where K
+   return (x.color == y.color && x.data == y.data)
+end
+
+Base.:(==)(x::RBTreeNode{K}, y::RBTreeNode{K}) where K = isequal(x, y)
+
+Base.hash(x::RBTreeNode) = hash(x.data, hash(x.color))
+
+function create_null_node(K::Type)
+    node = RBTreeNode{K}()
+    node.color = false
+    return node
+end 
+
+mutable struct RBTree{K}
+    root::RBTreeNode{K}
+    nil::RBTreeNode{K}
+
+    function RBTree{K}() where K 
+        new{K}(create_null_node(K), create_null_node(K))
+    end
+end
+
+RBTree() = RBTree{Any}()
+
+"""
+    search_node(tree, key)
+
+Returns the last visited node, while traversing through in binary-search-tree fashion looking for `key`.
+"""
+search_node(tree, key)
+
+function search_node(tree::RBTree{K}, d::K) where K
+    node = tree.root
+    while node != tree.nil && d != node.data
+        if d < node.data
+            node = node.leftChild
+        else
+            node = node.rightChild
+        end
+    end
+    return node
+end
+
+"""
+    search_key(tree, key)
+
+Returns true if `key` is present in the `tree`, else returns false.    
+"""
+search_key(tree, key)
+
+function search_key(tree::RBTree{K}, d::K) where K 
+    node = search_node(tree, d)
+    return (node.data == d)
+end
+
+"""
+    insert_node!(tree::RBTree, node::RBTreeNode)
+
+Inserts `node` at proper location by traversing through the `tree` in a binary-search-tree fashion.
+"""
+function insert_node!(tree::RBTree, node::RBTreeNode)
```

Comment:
Can we use this to provide `Base.push!(tree, value}` where `value` is some time type we can convert to `K`